File: max-points-on-a-line.cpp
/**
 * Definition for a point.
 * struct Point {
 *     int x;
 *     int y;
 *     Point() : x(0), y(0) {}
 *     Point(int a, int b) : x(a), y(b) {}
 * };
 */
#include <algorithm>
using namespace std;


class Solution {
public:
    bool cmp(Point x, Point y)
    {
        return x.x > y.x || (x.x == y.x && x.y > y.y);
    }
    
    bool OneLine(Point p1, Point p2, Point p3)
    {
        return (p2.y - p1.y) * (p3.x - p1.x) == (p3.y - p1.y)*(p2.x-p1.x);
    }

    int maxPoints(vector<Point> &points) {
        if (points.size() <= 2) return points.size();
        
        int cnt = points.size();
        for (int i = 0;i < cnt; i++)
            for (int j=0; j<cnt; j++)
            {
                if (cmp(points[i], points[j]))
                    swap(points[i], points[j]);
            }
        vector<int> flags;flags.push_back(1);
        for (int i=1; i<cnt; ++i)
            if (points[i].x == points[i-1].x && points[i].y==points[i-1].y)
                flags.push_back(0);
            else
                flags.push_back(1);
        vector<Point> _points;
        for(int i = 0;i < cnt; ++i)
        {
            if (flags[i])
                _points.push_back(points[i]);
        }
        cnt = _points.size();
        int ans = 0;
        for (int i = 0;i < cnt;i ++)
            for (int j = i + 1;j < cnt;j ++)
            {
                
                int cur = 2;
                for (int k=0; k<cnt && k!=i && k!=j; ++k)
                {
                    if (OneLine(_points[i], _points[j], _points[k]))
                        cur ++ ;
                }
                ans = max(ans, cur);
            }
        return ans;
    }
};

File: max-points-on-a-line.py
# Definition for a point
# class Point:
#     def __init__(self, a=0, b=0):
#         self.x = a
#         self.y = b

class Solution:
    
    def one_line(self, pnt_i, pnt_j, pnt_k):
        if pnt_i.x == pnt_j.x and pnt_j.x == pnt_k.x:
            return True;
        if pnt_i.y == pnt_j.y and pnt_j.y == pnt_k.y:
            return True;
        return (pnt_j.y - pnt_i.y) * (pnt_k.x - pnt_j.x) == (pnt_j.x - pnt_i.x)*(pnt_k.y - pnt_j.y)
        
    # @param points, a list of Points
    # @return an integer
    def maxPoints(self, points):
        cnt = dict()
        pnt = list()
        ans = 0
        for point in points:
            try:
                cnt[point] = cnt[point] + 1
                ans = max(ans, cnt[point])
            except KeyError as e:
                cnt[point] = 1
                pnt.append(point)
        print(ans)
        for i in range(len(pnt)):
            for j in range(i+1, len(pnt)):
                pnt_i = pnt[i]
                pnt_j = pnt[j]
                tmp = cnt[pnt_i] + cnt[pnt_j]
                
                for k in range(j+1, len(pnt)):
                    if self.one_line(pnt_i, pnt_j, pnt[k]):
                        tmp = tmp + cnt[pnt[k]]
                
                ans = max(ans, tmp)
        return ans;
